src/lib/crypto/signer.ts

```diff
@@ -0,0 +1,151 @@
+const secp256k1 = require('secp256k1')
+import * as utils from "../utils"
+
+import { Balance, Network, QueryStakesOrder, StakeEntry, UtxoMetadata } from "../types"
+import { IBIP32, IProvider, ISigner } from "./interfaces"
+import { KeyedSignature, PublicKey, PublicKeyHashString, UtxoSelectionStrategy } from "./types"
+
+export class Signer implements ISigner {
+    
+    protected node: IBIP32;
+    protected utxos: Array<UtxoMetadata> = []
+    
+    public readonly provider: IProvider;
+    public strategy: UtxoSelectionStrategy;
+
+    constructor(node: IBIP32, provider: IProvider, strategy?: UtxoSelectionStrategy) {
+        this.node = node
+        this.provider = provider
+        this.strategy = strategy || UtxoSelectionStrategy.SmallFirst
+        if (!provider.network) {
+            throw Error(`Signer: interal error: unintialized provider.`)
+        }
+    }
+
+    public get network(): Network {
+        return this.provider.network || "mainnet"
+    }
+
+    public get pkh(): string {
+        return this.publicKey.hash().toBech32(this.network)
+    }
+
+    public get publicKey(): PublicKey {
+        return PublicKey.fromUint8Array(this.node.publicKey)
+    }
+
+    public addUtxos(...utxos: Array<UtxoMetadata>) {
+        // avoid adding duplicates
+        const newUtxos: Array<UtxoMetadata> = []
+        utxos.forEach(utxo => {
+            if (this.utxos.findIndex(cached => cached.output_pointer === utxo.output_pointer) < 0) {
+                newUtxos.push(utxo)
+            }
+        })
```

Comment:
You can use Set to avoid duplicated
```suggestion
        const existingPointers = new Set(this.utxos.map(cached => cached.output_pointer));
        const newUtxos = utxos.filter(utxo => !existingPointers.has(utxo.output_pointer));
```